### PR TITLE
Loss model evaluation -> metric model evaluation

### PR DESCRIPTION
**Pull request by SaharaSurfer**

### CODE REVIEW

File: degann/networks/metrics.py [hunk 1]
```diff
@@ -44,9 +44,9 @@
 
 
 _metrics: dict = {
-    "RootMeanSquaredError": keras.metrics.RootMeanSquaredError(),
-    # "MaxDeviation": MaxDeviation(),
-    # "MeanDeviation": MeanDeviation(),
+    "root_mean_squared_error": keras.metrics.RootMeanSquaredError(),
+    # "max_deviation": MaxDeviation(),
+    # "mean_deviation": MeanDeviation(),
```
Comment (Krekep) [inline degann/networks/metrics.py:49]: Зачем переименовывать?

File: degann/search_algorithms/grid_search.py [hunk 2]
```diff
@@ -27,7 +27,7 @@ def grid_search_step(
     logging: bool = False,
     file_name: str = "",
     callbacks: Optional[list] = None,
-    metrics: Optional[list[str]] = None,
+    eval_metric: str = "root_mean_squared_error",
```
Comment (Krekep) [inline degann/search_algorithms/grid_search.py]: Помимо метрики, по которой идёт оценивание моделей, вполне естественно, если пользователь дополнительно хочет отслеживать состояние других (дополнительных метрик)

File: degann/search_algorithms/grid_search.py [hunk 3]
```diff
@@ -110,19 +106,18 @@ def grid_search_step(
                 epoch=num_epoch,
                 optimizer=opt,
                 loss_function=loss,
-                loss=curr_loss,
-                validation_loss=eval_loss,
+                metric_value=curr_metric_value,
+                validation_metric_value=val_metric_value,
```
Comment (Krekep) [inline degann/search_algorithms/grid_search.py]: Значение функции потерь всё равно стоит логгировать

File: degann/search_algorithms/pattern_search.py [hunk 4]
```diff
@@ -285,7 +285,7 @@ def pattern_search(
         map(
             lambda kv: (
                 [(kv[0], v) for v in kv[1]]
-                if isinstance(kv[1], Collection) and len(kv[1]) > 0
+                if isinstance(kv[1], list) and len(kv[1]) > 0
```
Comment (Krekep) [inline degann/search_algorithms/pattern_search.py:288]: Чем плох Collection?

File: degann/search_algorithms/pattern_search.py [hunk 5]
```diff
@@ -326,10 +326,7 @@ def __init__(self: "PatternSearchConfig") -> None:
             key for key in losses.get_all_loss_functions()
         ]
         self.optimizers: list[str] = [key for key in get_all_optimizers()]
-        self.metrics: list[list[str]] = [[key for key in get_all_metric_functions()]]
-        self.validation_metrics: list[list[str]] = [
-            [key for key in get_all_metric_functions()]
-        ]
+        self.eval_metric: str = "root_mean_squared_error"
```
Comment (Krekep) [inline degann/search_algorithms/pattern_search.py]: Тот же комментарий, что и выше. Доп. метрики нужны

File: degann/search_algorithms/random_search.py [hunk 6]
```diff
@@ -40,9 +40,9 @@ def random_search(
     if parameters.nn_alphabet is None:
         parameters.nn_alphabet = default_alphabet
 
-    best_net: dict
-    best_loss = 1e6
-    best_epoch: int
+    best_net = dict()
+    best_metric_value = 1e6
+    best_epoch = 0
```
Comment (Krekep) [inline degann/search_algorithms/random_search.py]: Зачем присваивать значения best_net и best_epoch?

File: degann/search_algorithms/random_search.py [hunk 7]
```diff
@@ -145,27 +147,33 @@ def random_search_endless(
...
-    nn_loss, nn_epoch, loss_f, opt_n, net = random_search(parameters)
+    nn_metric_value, nn_epoch, loss_f, opt_n, net = random_search(parameters)
     i = 1
     best_net = net
-    best_loss = nn_loss
+    best_metric_value = nn_metric_value
     best_epoch = nn_epoch
-    while nn_loss > parameters.loss_threshold and i != parameters.max_launches:
+    best_loss_f = loss_f
+    best_opt = opt_n
```
Comment (Krekep) [inline degann/search_algorithms/random_search.py]: А зачем best_loss_f и best_opt? Вроде нет поиска по алгоритму обучения и функции потерь?

File: degann/search_algorithms/search_algorithms_parameters.py [hunk 8]
```diff
@@ -92,7 +92,7 @@ def __init__(self) -> None:
         self.callbacks: Optional[list] = None
         self.logging: bool = False
         self.file_name: str = ""
-        self.metrics: Optional[list[str]] = None
+        self.eval_metric: str = "root_mean_squared_error"
```
Comment (Krekep) [inline degann/search_algorithms/search_algorithms_parameters.py]: Same

File: degann/search_algorithms/simulated_annealing.py [hunk 9]
```diff
@@ -179,13 +189,13 @@ def simulated_annealing(
                 epoch=gen_neighbor[1].value(),
                 optimizer=parameters.optimizer,
                 loss_function=parameters.loss_function,
-                loss=neighbor_loss,
-                validation_loss=neighbor_val_loss,
+                metric_value=neighbor_metric_value,
+                validation_metric_value=neighbor_val_metric_value,
```
Comment (Krekep) [inline degann/search_algorithms/simulated_annealing.py]: И тут про логгирование комментарий

File: tests/test_search_algorithms.py
Comment (Krekep) [inline tests/test_search_algorithms.py:1]: Добавьте тесты проверяющие корректность алгоритмов поиска. 
Типа того, что случайный поиск получает нейронку с метрикой ниже порога

**Review by Krekep — CHANGES_REQUESTED** [review]
Очень хорошо, но надо пошлифовать

**Comment by Krekep** [comment]
И ещё два замечания.
1. Добавляйте к PR описание того, что сделали (иногда бывает полезно добавлять ещё объяснение почему так сделано)
2. Не забывайте ревью запрашивать

File: degann/networks/metrics.py (hunk 1 above)
Comment (SaharaSurfer) [reply degann/networks/metrics.py:49]: Названия в словаре изменены на названия по умолчанию для самих метрик в keras, так как в этом случае метрику для оценки моделей очень легко найти в объекте History, который получаем в результате тренировки

File: degann/search_algorithms/pattern_search.py (hunk 4 above)
Comment (SaharaSurfer) [reply degann/search_algorithms/pattern_search.py:288]: Дело в том, что PatternSearchConfig все поля, кроме eval_metric представляют собой списки. Если мы используем в этом месте Collection (строка тоже считается), то от eval_metric ничего не останется, так как её разделит на отдельные буквы

File: degann/search_algorithms/random_search.py (hunk 6 above)
Comment (SaharaSurfer) [reply degann/search_algorithms/random_search.py]: Если я правильно понимаю, в классе для параметров случайного поиска нет никакого ограничения на iterations. Поэтому я решил сделать так, чтобы при iterations=0 не вылетала ошибка

File: degann/search_algorithms/random_search.py (hunk 7 above)
Comment (SaharaSurfer) [reply degann/search_algorithms/random_search.py]: Согласен, это лишнее

File: degann/search_algorithms/simulated_annealing.py (hunk 9 above)
Comment (SaharaSurfer) [reply degann/search_algorithms/simulated_annealing.py]: Логирование вернул

File: degann/search_algorithms/search_algorithms_parameters.py (hunk 8 above)
Comment (SaharaSurfer) [reply degann/search_algorithms/search_algorithms_parameters.py]: Вернул

File: degann/search_algorithms/pattern_search.py (hunk 5 above)
Comment (SaharaSurfer) [reply degann/search_algorithms/pattern_search.py]: Вернул

File: degann/search_algorithms/grid_search.py (hunk 3 above)
Comment (SaharaSurfer) [reply degann/search_algorithms/grid_search.py]: Логирование вернул

File: degann/search_algorithms/grid_search.py (hunk 2 above)
Comment (SaharaSurfer) [reply degann/search_algorithms/grid_search.py]: Вернул

File: tests/test_search_algorithms.py
Comment (SaharaSurfer) [reply tests/test_search_algorithms.py:1]: Добавил названный тест для random_search и sam

File: degann/networks/metrics.py (hunk 1 above)
Comment (Krekep) [reply degann/networks/metrics.py:49]: Мне казалось там можно указывать имена в PascalCase (как на скрине), но если нельзя, то окей, ничего критичного в переименовании нет
изображение

File: degann/search_algorithms/random_search.py (hunk 6 above)
Comment (Krekep) [reply degann/search_algorithms/random_search.py]: Лучше добавить проверку (assert к примеру) на то, что число итераций положительно

File: degann/search_algorithms/grid_search.py [hunk 10]
```diff
@@ -27,7 +27,8 @@ def grid_search_step(
     logging: bool = False,
     file_name: str = "",
     callbacks: Optional[list] = None,
-    metrics: Optional[list[str]] = None,
+    metrics: list[str] = [],
```
Comment (Krekep) [inline degann/search_algorithms/grid_search.py]: Метрики опциональны, лучше не использовать изменяемые объекты в качестве значений по умолчанию

File: degann/search_algorithms/pattern_search.py [hunk 11]
```diff
@@ -327,9 +327,7 @@ def __init__(self: "PatternSearchConfig") -> None:
         ]
         self.optimizers: list[str] = [key for key in get_all_optimizers()]
         self.metrics: list[list[str]] = [[key for key in get_all_metric_functions()]]
-        self.validation_metrics: list[list[str]] = [
-            [key for key in get_all_metric_functions()]
-        ]
```
Comment (Krekep) [inline degann/search_algorithms/pattern_search.py:332]: Зачем убирать метрики для валидации?

File: tests/test_search_algorithms.py [hunk 12]
```diff
@@ -191,19 +191,19 @@ def test_sam(equation_data, in_size, out_size):
     simulated_annealing_parameters.loss_function = "Huber"
     simulated_annealing_parameters.min_epoch = 5
     simulated_annealing_parameters.max_epoch = 10
-    simulated_annealing_parameters.loss_threshold = 1
+    simulated_annealing_parameters.metric_threshold = 5
```
Comment (Krekep) [inline tests/test_search_algorithms.py]: Тут лучше поставить значение метрики поменьше, т.к. `5` по метрике `rmse` это много, а тест должен проверять корректность алгоритма поиска (тоже самое и для теста случайного поиска)

**Review by Krekep — CHANGES_REQUESTED** [review]
Осталась пара небольших изменений и можно сливать

File: degann/networks/metrics.py (hunk 1 above)
Comment (SaharaSurfer) [reply degann/networks/metrics.py:49]: Это правда, но, как я понимаю, в `history = model.train(...)` метрики записаны в виде `"categorical_crossentropy"`. И чтобы потом без проблем доставать метрику для оценки модели, мне показалось, что лучше переписать все метрики в таком виде

File: degann/search_algorithms/pattern_search.py (hunk 11 above)
Comment (SaharaSurfer) [reply degann/search_algorithms/pattern_search.py:332]: Насколько я вижу, они нигде не используются (в `compile` передавались только `metrics`), да и я не совсем понимаю, почему не хватает обычных метрик

File: degann/search_algorithms/random_search.py (hunk 6 above)
Comment (SaharaSurfer) [reply degann/search_algorithms/random_search.py]: Добавил assert

File: degann/search_algorithms/grid_search.py (hunk 10 above)
Comment (SaharaSurfer) [reply degann/search_algorithms/grid_search.py]: Поправил

File: tests/test_search_algorithms.py (hunk 12 above)
Comment (SaharaSurfer) [reply tests/test_search_algorithms.py]: `metric_threshold` уполовинил, но поставил `max_launches = -1` в случайном поиске, иначе есть шанс, что алгоритм просто не успеет найти подходящую модель.

**Review by Krekep — APPROVED** [review]
Всё хорошо, спасибо за работу
